File: src/main/java/com/github/caijh/commons/util/DateUtils.java
package com.github.caijh.commons.util;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.util.Date;
import javax.annotation.Nonnull;

import com.github.caijh.commons.util.exception.DateException;
import com.github.caijh.commons.util.exception.InstanceNotSupportException;
import org.joda.time.Days;
import org.joda.time.LocalDate;
import org.joda.time.LocalDateTime;

/**
 * DateUtils.
 *
 * @author caijunhui
 * @since 2017/5/23
 */
public class DateUtils {

    private static final String DATE_TIME_FORMAT_DEFAULT = "yyyy-MM-dd HH:mm:ss";

    private DateUtils() {
        throw new InstanceNotSupportException();
    }

    public static Date now() {
        return new Date();
    }

    public static Date now(String pattern) {
        return parse(nowAsString(pattern), pattern);
    }

    public static String nowAsString() {
        return format(now());
    }

    public static String nowAsString(String pattern) {
        return format(now(), pattern);
    }

    public static long currentTimestamp() {
        return Instant.now().getEpochSecond();
    }

    public static long getTimestamp(Date date) {
        return date.toInstant().getEpochSecond();
    }

    public static String format(Date date) {
        return format(date, DATE_TIME_FORMAT_DEFAULT);
    }

    public static String format(Date date, String pattern) {
        return new SimpleDateFormat(pattern).format(date);
    }

    public static Date parse(String s) {
        return parse(s, DATE_TIME_FORMAT_DEFAULT);
    }

    public static Date parse(String s, String pattern) {
        try {
            return new SimpleDateFormat(pattern).parse(s);
        } catch (ParseException e) {
            throw new DateException(e);
        }
    }

    public static int daysBetween(@Nonnull Date d1, @Nonnull Date d2) {
        return Days.daysBetween(LocalDate.fromDateFields(d1), LocalDate.fromDateFields(d2)).getDays();
    }

    public static Date addSeconds(Date date, int seconds) {
        return LocalDateTime.fromDateFields(date).plusSeconds(seconds).toDate();
    }

    public static Date addDays(Date date, int days) {
        return LocalDateTime.fromDateFields(date).plusDays(days).toDate();
    }


}

File: src/main/kotlin/com/github/caijh/commons/util/DateUtils.kt
package com.github.caijh.commons.util

import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.*
import java.util.*


/**
 * DateUtils.
 *
 * @author caijunhui
 * @since 2017/5/23
 */
class DateUtils private constructor() {
    companion object {
        private const val DATE_TIME_FORMAT_DEFAULT = "yyyy-MM-dd HH:mm:ss"

        @JvmStatic
        fun now(): Date {
            return Date()
        }

        @JvmStatic
        fun now(pattern: String): Date {
            return parse(nowAsString(pattern), pattern)
        }

        @JvmStatic
        fun nowAsString(): String {
            return format(now())
        }

        @JvmStatic
        fun nowAsString(pattern: String): String {
            return format(now(), pattern)
        }

        @JvmStatic
        fun currentTimestamp(): Long {
            return Instant.now().epochSecond
        }

        @JvmStatic
        fun getTimestamp(date: Date): Long {
            return date.toInstant().epochSecond
        }

        @JvmStatic
        @JvmOverloads
        fun format(date: Date?, pattern: String = DATE_TIME_FORMAT_DEFAULT): String {
            if (date == null) {
                return ""
            }
            return SimpleDateFormat(pattern).format(date)
        }

        @JvmStatic
        @JvmOverloads
        fun parse(s: String?, pattern: String = DATE_TIME_FORMAT_DEFAULT): Date {
            return try {
                SimpleDateFormat(pattern).parse(s)
            } catch (e: ParseException) {
                throw DateException(e)
            }
        }

        @JvmStatic
        fun daysBetween(d1: Date, d2: Date): Int {
            return Period.between(asLocalDate(d1), asLocalDate(d2)).days
        }

        @JvmStatic
        fun addSeconds(date: Date, seconds: Int): Date {
            return Date.from(
                asLocalDateTime(date).plusSeconds(seconds.toLong()).atZone(ZoneId.systemDefault()).toInstant()
            )
        }

        @JvmStatic
        fun addDays(date: Date, days: Int): Date {
            return Date.from(asLocalDateTime(date).plusDays(days.toLong()).atZone(ZoneId.systemDefault()).toInstant())
        }

        @JvmStatic
        fun asLocalDate(date: Date): LocalDate? {
            return Instant.ofEpochMilli(date.time).atZone(ZoneId.systemDefault()).toLocalDate()
        }

        @JvmStatic
        fun asLocalDateTime(date: Date): LocalDateTime {
            return Instant.ofEpochMilli(date.time).atZone(ZoneId.systemDefault()).toLocalDateTime()
        }

        @JvmStatic
        fun asDate(localDate: LocalDate): Date {
            return Date.from(localDate.atStartOfDay().atZone(ZoneId.systemDefault()).toInstant())
        }

        @JvmStatic
        fun asDate(localDateTime: LocalDateTime): Date {
            return Date.from(localDateTime.atZone(ZoneId.systemDefault()).toInstant())
        }
    }

    init {
        throw InstanceNotSupportException()
    }
}
